presence_channel.go

```diff
@@ -0,0 +1,330 @@
+package pusher
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var (
+	// ErrNotSubscribed is returned by functions that require the channel to be
+	// subscribed before being called.
+	ErrNotSubscribed = errors.New("not subscribed")
+	// ErrMissingMe indicates a presence channel was subscribed to, but the pusher
+	// server violated the message protocol by not providing a member for the
+	// current user.
+	ErrMissingMe = errors.New("missing member for current user")
+)
+
+// Member represents a channel member.
+type Member struct {
+	ID string
+	// Info is the JSON set by the auth server. Do not modify the underlying byte
+	// slice, it's shared by all instances of the member.
+	Info json.RawMessage
+}
+
+// PresenceChannel provides information about the users that are currently
+// subscribed.
+//
+// Note: Bind() does not fire pusher:member_added/removed, use
+// BindMemberAdded/Removed() instead.
+type PresenceChannel interface {
+	Channel
+
+	// BindMemberAdded returns a channel that receives a Member value when a user
+	// joins the channel. Events may be delivered out of order. Use
+	// UnbindMemberAdded when finished listening to events.
+	BindMemberAdded() chan Member
+
+	// UnbindMemberAdded removes bindings created by BindMemberAdded(). If chans
+	// are passed, only those bindings will be removed. Otherwise, all bindings
+	// for this event will be removed.
+	UnbindMemberAdded(...chan Member)
+
+	// BindMemberRemoved returns a channel that receives a user ID when a user
+	// leaves the channel. Events may be delivered out of order. Use
+	// UnbindMemberRemoved when finished listening for events.
+	BindMemberRemoved() chan string
+
+	// UnbindMemberRemoved removes bindings created by UnbindMemberRemoved(). If
+	// chans are passed, only those bindings will be removed. Otherwise, all
+	// bindings for this event will be removed.
+	UnbindMemberRemoved(...chan string)
+
+	// Members returns the member info of each user currently subscribed to the
+	// channel.
+	Members() map[string]Member
+
+	// Member returns a representation of the channel member with the given ID.
+	// `nil` is returned if the member isn't in the channel.
+	Member(id string) *Member
+
+	// Me returns the member for the current user.
+	//
+	// Possible errors:
+	//  - not subscribed - subscription must succeed before calling Me()
+	//  - invalid channel data - pusher server provided invalid JSON
+	//  - missing member for current user - pusher server violated protocol
+	Me() (*Member, error)
+
+	// MemberCount returns the number of users connected to the channel.
+	MemberCount() int
+}
+
+// presenceChannel implements the internalChannel and PresenceChannel interfaces
+type presenceChannel struct {
+	*privateChannel
+
+	memberAddedChans   map[chan Member]chanContext
+	memberRemovedChans map[chan string]chanContext
+	members            map[string]Member
+
+	membersMutex sync.RWMutex
```

Comment:
nit: usually the convention is to place mutex in front of fields it guards.
```suggestion
	membersMutex       sync.RWMutex
	memberAddedChans   map[chan Member]chanContext
	memberRemovedChans map[chan string]chanContext
	members            map[string]Member
```